pkg/helios/cas/cas_test.go

```diff
@@ -0,0 +1,499 @@
+		for i, content := range contents {
+			hash, err := store.Store(content)
+			require.NoError(t, err)
+			hashes[i] = hash
+		}
+		batchStoreDuration := time.Since(start)
+		
+		// Average per-operation should still be fast
+		avgStoreTime := batchStoreDuration / time.Duration(batchSize)
+		assert.Less(t, avgStoreTime, 1*time.Millisecond,
+			"Average store time in batch: %v, should be <1ms", avgStoreTime)
+		
+		// Batch load operations
+		start = time.Now()
+		for i, hash := range hashes {
+			retrieved, err := store.Load(hash)
+			require.NoError(t, err)
+			assert.Equal(t, contents[i], retrieved)
+		}
+		batchLoadDuration := time.Since(start)
+		
+		avgLoadTime := batchLoadDuration / time.Duration(batchSize)
+		assert.Less(t, avgLoadTime, 5*time.Millisecond,
+			"Average load time in batch: %v, should be <5ms", avgLoadTime)
+	})
+
+	t.Run("Memory_Usage_Efficiency", func(t *testing.T) {
+		// Test that duplicate content doesn't consume extra storage
+		content := []byte("duplicate content test")
+		duplicateCount := 1000
+		
+		// Store the same content multiple times
+		var firstHash types.Hash
+		for i := 0; i < duplicateCount; i++ {
+			hash, err := store.Store(content)
+			require.NoError(t, err)
+			
+			if i == 0 {
+				firstHash = hash
+			} else {
+				assert.Equal(t, firstHash, hash, "Hash should be identical for duplicate content")
+			}
+		}
+		
+		// Verify content is still retrievable
+		retrieved, err := store.Load(firstHash)
+		require.NoError(t, err)
+		assert.Equal(t, content, retrieved)
+	})
+}
+
+// TestContentAddressableStorage_ErrorHandling tests error conditions
+func TestContentAddressableStorage_ErrorHandling(t *testing.T) {
+	tempDir := t.TempDir()
+	store, err := NewBLAKE3Store(tempDir)
+	require.NoError(t, err)
+	defer store.Close()
+
+	t.Run("Load_NonExistent_Hash", func(t *testing.T) {
+		nonExistentHash := types.Hash{
+			Algorithm: types.BLAKE3,
+			Digest:    bytes.Repeat([]byte{0xFF}, 32), // Unlikely to exist
+		}
+		
+		data, err := store.Load(nonExistentHash)
+		assert.Error(t, err)
+		assert.Nil(t, data)
+		assert.Contains(t, err.Error(), "not found") // Should be descriptive
+	})
+
+	t.Run("Store_Empty_Content", func(t *testing.T) {
+		// Empty content should still be hashable and storable
+		emptyContent := []byte{}
+		
+		hash, err := store.Store(emptyContent)
+		require.NoError(t, err)
+		assert.NotEmpty(t, hash.Digest)
+		
+		retrieved, err := store.Load(hash)
+		require.NoError(t, err)
+		assert.Equal(t, emptyContent, retrieved)
+	})
+
+	t.Run("Invalid_Hash_Algorithm", func(t *testing.T) {
+		invalidHash := types.Hash{
+			Algorithm: "invalid-algorithm",
+			Digest:    make([]byte, 32),
+		}
+		
+		data, err := store.Load(invalidHash)
+		assert.Error(t, err)
+		assert.Nil(t, data)
+		assert.Contains(t, err.Error(), "unsupported hash algorithm")
+	})
+}
+
+// TestContentAddressableStorage_Concurrency tests concurrent operations
+func TestContentAddressableStorage_Concurrency(t *testing.T) {
```

Comment:
Tests: add a shutdown-concurrency test that enqueues Stores while calling Close(), asserting no panic (send on closed channel) and timely shutdown. Running with -race in CI would also catch the s.closed data race.